### PR TITLE
Equihash: Pass each obtained solution to a callback for immediate checking

### LINKED ISSUE #1143: Equihash: Return valid solutions immediately
Currently, the Equihash solver collects all valid solutions from a solver run and then returns them as a vector. We should instead pass in a callback to the solver to check obtained solutions against the difficulty immediately (the same way we have the `cancelled` callback).

### PR DESCRIPTION
**Pull request by str4d**
Closes #1143

### CODE REVIEW

**Comment by daira** [comment]
ACK.

**Comment by str4d** [comment]
@zkbot r+

**Comment by str4d** [comment]
Buildbot is hanging on RPC tests because of a bug in this commit.

File: src/rpcmining.cpp [hunk 1]
```diff
@@ -188,19 +188,13 @@ Value generate(const Array& params, bool fHelp)
...
+                    [&pblock](std::vector<eh_index> soln) {
                 pblock->nSolution = soln;
-
-                if (CheckProofOfWork(pblock->GetHash(), pblock->nBits, Params().GetConsensus())) {
-                    goto endloop;
-                }
-            }
+                return CheckProofOfWork(pblock->GetHash(), pblock->nBits, Params().GetConsensus());
+            };
+            if (EhBasicSolveUncancellable(n, k, curr_state, validBlock));
```
Comment (str4d) [inline src/rpcmining.cpp]: Can you spot the problem with this line? :mag_right:

This is a good example of an easily-overlooked change that can subtly change the execution logic. Or not-so-subtly in this case, thanks to the tests!

**Comment by ebfull** [comment]
@zkbot r+ 4a0bdc90200e63926029357e304cd0fa7716d61e

**Comment by ebfull** [comment]
Yes, let's make brackets a requirement on all code. I wish we could lint it.

**Pull request merged**